### PR TITLE
Fix build on macOS (10.x / 11.x)

### PR DESCRIPTION
**Pull request by 3wnbr1**
---

## Basic Info

macOS : 11.0.1 (20B5022a)
Build tools : Apple Clang 12.0.x
ROS2 : Foxy (Patch release 3)

---

## Description of contribution in a few bullet points

Building navigation2 on latest macOS 11 (Big Sur) is now possible with a ros2 foxy environment. Previous macOS versions should also work but remains untested. This PR adresses the following fatal build errors (with tests set to OFF) :

#### nav2_util
```
--- stderr: nav2_util
/Users/ewen/ros_ws/navigation2/nav2_util/src/dump_params.cpp:351:33: error: use of undeclared identifier 'basename'
      std::cout << "Usage: " << basename(argv[0]) << "\n";
```

#### nav2_amcl
```
--- stderr: nav2_amcl
/Users/ewen/ros_ws/navigation2/nav2_amcl/src/amcl_node.cpp:404:64: error: cannot pass object of non-trivial type 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char> >') through variadic function; call will abort at runtime [-Wnon-pod-varargs]
      " Verify that data is being published on the %s topic.", scan_topic_);
                                                               ^
/Users/ewen/ros_ws/navigation2/nav2_amcl/src/amcl_node.cpp:413:31: error: cannot pass object of non-trivial type 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char> >') through variadic function; call will abort at runtime [-Wnon-pod-varargs]
      d.nanoseconds() * 1e-9, scan_topic_);
```

#### nav2_behavior_tree
```
--- stderr: nav2_behavior_tree
In file included from /Users/ewen/ros_ws/navigation2/nav2_behavior_tree/plugins/action/reinitialize_global_localization_service.cpp:16:
In file included from /Users/ewen/ros_ws/navigation2/nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/reinitialize_global_localization_service.hpp:20:
/Users/ewen/ros_ws/navigation2/nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:38:34: error: use 'template' keyword to treat 'get' as a dependent template name
    node_ = config().blackboard->get<rclcpp::Node::SharedPtr>("node");
                                 ^
                                 template
/Users/ewen/ros_ws/navigation2/nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:42:28: error: use 'template' keyword to treat 'get' as a dependent template name
      config().blackboard->get<std::chrono::milliseconds>("server_timeout");
                           ^
                           template
/Users/ewen/ros_ws/navigation2/nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:129:26: error: use 'template' keyword to treat 'get' as a dependent template name
    config().blackboard->get<int>("number_recoveries", recovery_count);  // NOLINT
                         ^
                         template
/Users/ewen/ros_ws/navigation2/nav2_behavior_tree/include/nav2_behavior_tree/bt_service_node.hpp:131:26: error: use 'template' keyword to treat 'set' as a dependent template name
    config().blackboard->set<int>("number_recoveries", recovery_count);  // NOLINT
                         ^
                         template
```

#### nav2_costmap_2d
```
--- stderr: nav2_costmap_2d
In file included from /Users/ewen/ros_ws/navigation2/nav2_costmap_2d/plugins/obstacle_layer.cpp:39:
In file included from /Users/ewen/ros_ws/navigation2/nav2_costmap_2d/include/nav2_costmap_2d/obstacle_layer.hpp:58:
In file included from /Users/ewen/ros_ws/navigation2/nav2_costmap_2d/include/nav2_costmap_2d/observation_buffer.hpp:49:
/Users/ewen/ros_ws/navigation2/nav2_costmap_2d/include/nav2_costmap_2d/observation.hpp:57:11: error: definition of implicit copy assignment operator for 'Observation' is deprecated because it has a user-declared destructor [-Werror,-Wdeprecated]
  virtual ~Observation()
          ^
```

#### nav2_bt_navigator
```
--- stderr: nav2_bt_navigator
/Users/ewen/ros_ws/navigation2/nav2_bt_navigator/src/ros_topic_logger.cpp:44:5: error: no matching function for call to 'toMsg'
    tf2_ros::toMsg(std::chrono::time_point<std::chrono::high_resolution_clock>(timestamp));
    ^~~~~~~~~~~~~~
/Users/ewen/ros2_foxy/install/tf2_ros/include/tf2_ros/buffer_interface.h:48:40: note: candidate function not viable: no known conversion from 'time_point<std::chrono::high_resolution_clock, [...]>' to 'const time_point<std::chrono::system_clock, [...]>' for 1st argument
  inline builtin_interfaces::msg::Time toMsg(const tf2::TimePoint & t)
                                       ^
/Users/ewen/ros2_foxy/install/tf2_ros/include/tf2_ros/buffer_interface.h:67:44: note: candidate function not viable: no known conversion from 'std::chrono::time_point<std::chrono::high_resolution_clock>' (aka 'time_point<std::__1::chrono::steady_clock>') to 'const tf2::Duration' (aka 'const duration<long long, ratio<1LL, 1000000000LL> >') for 1st argument
  inline builtin_interfaces::msg::Duration toMsg(const tf2::Duration & t)
```

#### smac_planner
```
--- stderr: smac_planner
/Users/ewen/ros_ws/navigation2/smac_planner/src/a_star.cpp:355:19: error: constexpr variable 'sqrt_2' must be initialized by a constant expression
  constexpr float sqrt_2 = std::sqrt(2.);
                  ^        ~~~~~~~~~~~~~
/Users/ewen/ros_ws/navigation2/smac_planner/src/a_star.cpp:355:28: note: non-constexpr function 'sqrt' cannot be used in a constant expression
  constexpr float sqrt_2 = std::sqrt(2.);
```

## Description of documentation updates required from your changes

(Optional) Add instructions for installing dependencies and building for macOS.

```bash
# Install dependencies with Homebrew
brew install ompl

# Create build workspace
mkdir -p ~/ros_ws/src
cd ~/ros_ws/src
git clone https://github.com/BehaviorTree/BehaviorTree.CPP.git
git clone -b ros2 https://github.com/ros/angles.git
git clone -b foxy-devel https://github.com/ros-planning/navigation2.git

# Build without tests
colcon build --symlink-install --cmake-args "-DBUILD_TESTING=OFF"
```
---

## Future work that may be required in bullet points

Fix warnings in smac_planner 

```
In file included from /Users/ewen/ros_ws/navigation2/smac_planner/src/smac_planner_2d.cpp:21:
In file included from /Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/smac_planner_2d.hpp:23:
In file included from /Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/smoother.hpp:26:
/Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/smoother_cost_function.hpp:139:24: warning: using the result of an assignment as a condition without
      parentheses [-Wparentheses]
      if (valid_coords = _costmap->worldToMap(xi[0], xi[1], mx, my)) {
          ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/smoother_cost_function.hpp:139:24: note: place parentheses around the assignment to silence this warning
      if (valid_coords = _costmap->worldToMap(xi[0], xi[1], mx, my)) {
                       ^
          (                                                        )
/Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/smoother_cost_function.hpp:139:24: note: use '==' to turn this assignment into an equality comparison
      if (valid_coords = _costmap->worldToMap(xi[0], xi[1], mx, my)) {
                       ^
                       ==
1 warning generated.

/Users/ewen/ros_ws/navigation2/smac_planner/src/a_star.cpp:433:19: warning: unused parameter 'node' [-Wunused-parameter]
  const NodePtr & node,
                  ^
/Users/ewen/ros_ws/navigation2/smac_planner/src/a_star.cpp:434:22: warning: unused parameter 'node_getter' [-Wunused-parameter]
  const NodeGetter & node_getter)
                     ^

In file included from /Users/ewen/ros_ws/navigation2/smac_planner/src/a_star.cpp:33:
In file included from /Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/a_star.hpp:30:
/Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/node_se2.hpp:178:3: warning: definition of implicit copy constructor for 'NodeSE2' is deprecated because it has
      a user-declared destructor [-Wdeprecated]
  ~NodeSE2();
  ^

In file included from /Users/ewen/ros_ws/navigation2/smac_planner/src/a_star.cpp:33:
In file included from /Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/a_star.hpp:29:
/Users/ewen/ros_ws/navigation2/smac_planner/include/smac_planner/node_2d.hpp:69:3: warning: definition of implicit copy constructor for 'Node2D' is deprecated because it has a
      user-declared destructor [-Wdeprecated]
  ~Node2D();
  ^

### CODE REVIEW

**Review by SteveMacenski — COMMENTED** [review]
This looks good to me except https://github.com/ros-planning/navigation2/pull/2083/files#diff-8e4ad45054e440b04bf332cf96752ac6ae3d5b236829ccd2b593570c64a80c62L234 where I Don't think that fits the Foxy API

Also, revert the `sqrt(2)` in SMAC, don't approximate.

**Comment by SteveMacenski** [comment]
We don't accept PRs directly to Foxy, please rebase to master and I'll cherry pick over to foxy once merged.

**Comment by 3wnbr1** [comment]
Few changes to discuss there :


- A `constexpr` version of `std::sqrt`is not part of any C++ standard and hence doesn't exist in macOS libc (and probably many other platforms). [See this topic](https://stackoverflow.com/questions/8622256/in-c11-is-sqrt-defined-as-constexpr). The details are attached bellow.
```
/Users/ewen/ros_rol_ws/src/navigation2/smac_planner/src/a_star.cpp:355:19: error: constexpr variable 'sqrt_2' must be initialized by a constant expression
  constexpr float sqrt_2 = std::sqrt(2.);
                  ^        ~~~~~~~~~~~~~
/Users/ewen/ros_rol_ws/src/navigation2/smac_planner/src/a_star.cpp:355:28: note: non-constexpr function 'sqrt' cannot be used in a constant expression
  constexpr float sqrt_2 = std::sqrt(2.);
```

To circumvent this issue, the value of sqrt can be either hardcoded without precision loss (easy) or the code would require to be refactored.
```python
>>> import numpy as np
>>> np.sqrt(np.float128(2.))
1.4142135623730950488
```

- The standard upgrade C++17, bringing `std::filesystem` as `std::experimental::filesystem` is now considered deprecated by LLVM/Clang

Waiting for your feedback

**Comment by SteveMacenski** [comment]
I don’t allow for hardcoded values in the codebase. While this is obvious what it represents, magic numbers have a way of being forgotten over time what they are so as a general rule I don’t allow them unless in a mathematical expression with links to appropriate documentation. 

You could make it static const to get around it.

File: nav2_amcl/CMakeLists.txt
Comment (SteveMacenski) [inline nav2_amcl/CMakeLists.txt:5]: Revert to c++14. ROS2 doesnt support c++17 yet and should only be used when specifically required until that migration is made.

File: nav2_behavior_tree/CMakeLists.txt
Comment (SteveMacenski) [inline nav2_behavior_tree/CMakeLists.txt:5]: Revert all c++17 changes.

File: nav2_behavior_tree/CMakeLists.txt
Comment (SteveMacenski) [inline nav2_behavior_tree/CMakeLists.txt:9]: Revert. Our compile flags are set in nav2 common.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by 3wnbr1** [comment]
I didn't know about C++14 only for ROS2, changes were made as requested. I had to add an additional pragma to force LLVM to ignore deprecated declaration for `std::experimental::filesystem`

File: nav2_bt_navigator/src/ros_topic_logger.cpp
Comment (SteveMacenski) [inline nav2_bt_navigator/src/ros_topic_logger.cpp:45]: Is there really no way we can represent a timestamp that all of these can use? There are plenty of places in the codebase we use timestamps, but this is the only one with this level of if/elsing on operating systems. If we change `timestamp` to just a `double`, is there no way that all of these can handle it?

File: nav2_costmap_2d/include/nav2_costmap_2d/observation_buffer.hpp
Comment (SteveMacenski) [inline nav2_costmap_2d/include/nav2_costmap_2d/observation_buffer.hpp:56]: Rather than ignoring an unused variable, can we use it in the tf2 buffer calls or delete it?

File: nav2_waypoint_follower/include/nav2_waypoint_follower/plugins/photo_at_waypoint.hpp
Comment (SteveMacenski) [inline nav2_waypoint_follower/include/nav2_waypoint_follower/plugins/photo_at_waypoint.hpp:38]: Only use pragma's before and after specific lines you want to ignore, do not ignore entire files.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
Looks good, just some cleanup, then I can cherry pick to foxy

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**
